Add documentation for Boogie command line options (#2203)
dafny-lang/dafny: docs/DafnyRef/UserGuide.md

  seed. If not, it will implicitly use `-randomSeed:0` to ensure a
  difference between iterations. This option can be very useful for
  identifying input programs for which verification is unstable. If the
  verification times or solver resource counts associated with each
  proof attempt vary widely for a given procedure, small changes to that
  procedure might be more likely to cause proofs to fail in the future.

* `-vcsSplitOnEveryAssert` - prove each (explicit or implicit) assertion
  in each procedure separately. See also the attribute [`{:vcs_split_on_every_assert}`](#sec-vcs_split_on_every_assert) for restricting this option on specific procedures. By default, Boogie attempts to prove
  that every assertion in a given procedure holds all at once, in a
  single query to an SMT solver. This usually performs well, but
  sometimes causes the solver to take longer. If a proof that you
  believe should succeed is timing out, using this option can sometimes
  help.

* `-vcsCores:<n>` - try to verify `<n>` procedures simultaneously.
  Setting `<n>` to the number of physical cores available tends to be
  effective at speeding up overall proof time.

> * `-timeLimit:<n>` - spend at most `<n>` seconds attempting to prove any

Reviewer: The Dafny Attributes.md file mentions "Boogie `timelimit` command-line option"

Is it -timelimit or -timeLimit? Please fix the line that is wrong.

Author: Done.

  code, regardless of whether errors are found. If `1` (default) then
  use the appropriate exit code. This option is deprecated.

### 24.10.8. Controlling Boogie {#sec-controlling-boogie}

Dafny builds on top of Boogie, a general-purpose intermediate language
for verification. Options supported by Boogie on its own are also
supported by Dafny. Some of the Boogie options most relevant to Dafny
users include the following. We use the term "procedure" below to refer
to a Dafny function, lemma, method, or predicate, following Boogie
terminology.

> * `-proc:<name>` - verify only the procedure named `<name>`. The name

Reviewer 2: Wow I can't believe this wasn't documented before! Well done!

  itself.

  Certain Boogie inputs are unstable in the sense that changes to the
  input that preserve its meaning may cause the output to change. The
  `-randomSeed`` option simulates meaning-preserving changes to the
  input without requiring the user to actually make those changes.

  The `-randomSeed` option is implemented by renaming variables and
  reordering declarations in the input, and by setting
  solver options that have similar effects.

* `-randomSeedIterations:<n>` - attempt to prove each VC `<n>` times
  with `<n>` random seeds. If `-randomSeed` has been provided, each
  proof attempt will use a new random seed derived from this original
  seed. If not, it will implicitly use `-randomSeed:0` to ensure a
  difference between iterations. This option can be very useful for
  identifying input programs for which verification is unstable. If the
  verification times or solver resource counts associated with each
  proof attempt vary widely for a given procedure, small changes to that
  procedure might be more likely to cause proofs to fail in the future.

* `-vcsSplitOnEveryAssert` - prove each (explicit or implicit) assertion

>   in each procedure separately. By default, Boogie attempts to prove

Reviewer 2: ```suggestion
  in each procedure separately. See also the attribute [`{:vcs_split_on_every_assert}`](#sec-vcs_split_on_every_assert) for restricting this option on specific procedures. By default, Boogie attempts to prove
```
```suggestion
  in each procedure separately. By default, Boogie attempts to prove
```

  that every assertion in a given procedure holds all at once, in a
  single query to an SMT solver. This usually performs well, but
  sometimes causes the solver to take longer. If a proof that you
  believe should succeed is timing out, using this option can sometimes
  help.

* `-vcsCores:<n>` - try to verify `<n>` procedures simultaneously.
  Setting `<n>` to the number of physical cores available tends to be
  effective at speeding up overall proof time.

* `-timeLimit:<n>` - spend at most `<n>` seconds attempting to prove any

>   single SMT query.

Reviewer 2: ```suggestion
  single SMT query. This setting can also be set per method using the attribute [`{:timeLimit n}`](#sec-time-limit).
```
Also, can you please ensure the link is valid? I think you'll have to insert `{#sec-time-limit}` at the right place in `Attributes.md`

* `-rlimit:<n>` - set the maximum solver resource count to use while
  proving a single SMT query. This can be a more deterministic approach

>   than setting a time limit. To choose an appropriate value, try proving
>   some representative code and comparing the time and resource count
>   values reported to calculate a correspondence.

Reviewer 2: ```suggestion
  than setting a time limit. To choose an appropriate value, please refer to the documentation of the attribute [`{:rlimit}`](#sec-rlimit) that can be applied per procedure.
```

  effective at speeding up overall proof time.

* `-timeLimit:<n>` - spend at most `<n>` seconds attempting to prove any
  single SMT query. This setting can also be set per method using the attribute [`{:timeLimit n}`](#sec-time-limit).

> * `-rlimit:<n>` - set the maximum solver resource count to use while

Reviewer 2: I think this sentence needs a full stop

Author: It has one, though, doesn't it?
